keras_nlp/utils/text_generation.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Text generation utilities."""
+
+import tensorflow as tf
+
+
+def _handle_end_token(next_token, end_token_received, end_token_id):
+    filtered_next_token = next_token * (1 - end_token_received)
+    end_token_received = tf.cast(
+        filtered_next_token == end_token_id, dtype=next_token.dtype
+    )
+    return filtered_next_token, end_token_received
+
+
+def generate_text_greedy(
+    get_next_token_probability_fn, seed, max_length, end_token_id=None
+):
+    """
+    Text generation utility based on greedy search.
+
+    Greedy search always appends the token having the largest probability to
+    existing sequence.
+
+    Args:
+        get_next_token_probability_fn: a callable, which takes in input_sequence
+            and output the probability distribution of the next token.
+        seed: a list, the initial tokens to append generated tokens.
+        max_length: int. The max length of generated text.
+        end_token_id: int, defaults to None. The token marking the end of the
+            sequence, once encountered the generation is finished for the exact
+            sequence. If None, every sequence is generated to `max_length`.
+
+    Returns:
+        A 1D int Tensor, or 2D int Tensor of shape (batch_size, `max_length`)
+        representing the generated sequences.
+
+    Examples:
+    ```python
+    VOCAB_SIZE = 10
+    FEATURE_SIZE = 16
+
+    # Create a dummy model to predict the next token.
+    model = tf.keras.Sequential([
+        tf.keras.Input(shape=[None]),
+        tf.keras.layers.Embedding(
+            input_dim=VOCAB_SIZE,
+            output_dim=FEATURE_SIZE,),
+        tf.keras.layers.Dense(VOCAB_SIZE),
+        tf.keras.layers.Softmax(),
+    ])
+
+    # Define a function that outputs the next token's probability given the
+    # input sequence.
+    def get_next_token_probability_fn(inputs):
+        return model(inputs)[:, -1, :]
+
+    inputs = tf.random.uniform(shape=[5, 5], maxval=VOCAB_SIZE, dtype=tf.int64)
+
+    # Print the generated sequence (token ids).
+    print(generate_text_greedy(
+        get_next_token_probability_fn,
+        inputs,
+        max_length=10,
+        end_token=0,))
+    ```
+
+    """
+    if 0 in seed.shape:
```

Comment:
Are we sure we need to have this as a requirement?

For a decoder only model, if you aren't training with start and end markers, couldn't you just ask for a random prediction at position zero without any input?

We should make sure we aren't ruling out any valid workflow with this.

Reply:
We can always prompt our users to provide the starting sequence, if they want to use a random starting token, then they can construct it by themselves? Basically I want users to have the full control over generation, if they do not provide any inputs, then we will let them know.

Reply:
But they might want to sample the random starting token the same as the rest of the sequence is sampled right? Or include it in the beam search (when we implement that one).

I worry we are being too restrictive with this code requirement, but might need to play in code around a bit to develop this thought.